etcd/etcd.go

```diff
@@ -175,23 +175,25 @@ func isTimeout(err error) bool {
 // are tried. It returns the response to the first successful request.
 func (c *Client) tryEndpoints(mreq func(string) (*http.Request, error)) (*http.Response, error) {
 	var (
-		endpoints    []string
 		req          *http.Request
 		rsp          *http.Response
 		err          error
 		endpointErrs []error
 	)
 
-	endpoints = c.endpoints
-	for len(endpoints) > 0 {
-		req, err = mreq(endpoints[0] + "/v2/keys")
+	for index, endpoint := range c.endpoints {
+		req, err = mreq(endpoint + "/v2/keys")
 		if err != nil {
 			return nil, err
 		}
 
 		rsp, err = c.client.Do(req)
 		if err == nil {
-			break
+			if index != 0 {
+				c.endpoints = append(c.endpoints[index:], c.endpoints[:index]...)
+			}
+
+			return rsp, nil
```

Comment:
i like this. however, the same should be done to the timeout cases: return nil, err instead of break. then it would be cool.